dataprofiler/data_readers/csv_data.py

```diff
@@ -103,24 +103,152 @@ def _check_and_return_options(self, options):
                                  'integer for the row that represents the '
                                  'header (0 based index)')
         return options
-
+    
     @staticmethod
-    def _determine_delimiter_of_str(data_as_str):
+    def _guess_delimiter(data_as_str, preferred=[',', '\t'], omitted=['"', "'"]):
         """
         Automatically checks for what delimiter exists in a text document.
 
         :param data_as_str: Single string containing rows (lines separated by "\n")
         :type data_as_str: str
+        :param preferred: Delimiters that will be weighted and selected if possible
+        :type preferred: list
+        :param omitted: Delimiters that will be omitted from selection
+        :type omitted: list
         :return: Delimiter, if none can be found None is returned
         :rtype: str or None
         """
-        sniffer = csv.Sniffer()
-        sniffer.preferred = [',', '\t', ';']  # removes ' ', ':' from preferred
-        try:
-            dialect = sniffer.sniff(data_as_str)
-        except csv.Error as exc:
-            return None
-        return dialect.delimiter
+
+        # Detect vocabulary (and count)
+        vocab = {}
+        for i in range(len(data_as_str)):
+            c = data_as_str[i]
+            if c not in vocab:
+                vocab[c] = 0
+            vocab[c] += 1            
+        if '\n' in vocab: vocab.pop('\n')        
+        for char in omitted:
+            if char in vocab:
+                vocab.pop(char)
+
+        # Sort vocabulary by count
+        ordered_vocab = []
+        sorted_keys = sorted(vocab, key=vocab.get)
+        sorted_keys.reverse()
+        for c in sorted_keys:
+            ordered_vocab.append(c)
+
+        # Evaluate vocab, reviewing rows and columns
+        delimiter = None
+        validated_proposed_delimiters = {}
+        for proposed_delim in ordered_vocab[0:12]:
```

Comment:
Should we look into maybe making the upper limit here relative to the length of data_as_a_str or vocab? Like what if there are very few, but very long cells, and many characters appear more than twice, then they would appear more than the delimiter? Just a thought/discussion, a pretty edge one at that.

Reply:
Say we have 1000 columns, with 1000 words per column. 

The vocab is stored in a dictionary and a count is kept per character. So it's bound by the number of unique characters. Then I truncate to the top 12. How was 12 determined? The idea was 10 common digits + two special characters for eval (`0-9`, `' '`, `,`). 

Worked on all the files 🤷‍♂️

Reply:
Oh I suppose thats a good point, alright then.

Reply:
can we make it work for multiple-character delimiters?